### PR TITLE
Travis - Only run tests that have changed

### PR DESCRIPTION
**Pull request by jerjou**
(work in progress. Will assign when ready)

### CODE REVIEW

**Comment by jerjou** [comment]
Okay - 'tis ready.

File: travis.sh [hunk 1]
```diff
@@ -23,8 +23,44 @@ SKIP_TESTS=false
...
+
+    changed="$(git diff --name-only "$TRAVIS_COMMIT" "$TRAVIS_BRANCH")"
+    if [ $? -ne 0 ]; then
+      # Fall back to git head
+      changed="$(git diff --name-only "$(git rev-parse HEAD)" "$TRAVIS_BRANCH")"
+      [ $? -ne 0 ] && return 0  # Give up. Just run everything.
+    fi
+
+    # Find the common prefix
+    prefix="$(echo "$changed" | grep -v travis.sh | sed -e 'N;s/^\(.*\).*\n\1.*$/\1\n\1/;D')"
```
Comment (tswast) [inline travis.sh]: Magic! I might want a bit more explanation in the comments for the sed command. (Something like https://robots.thoughtbot.com/literate-vim but maybe not quite that detailed)

**Comment by tswast** [comment]
@lesv please take a look, too.

**Comment by jerjou** [comment]
@tswast Added some comments. LMKWYT

File: travis.sh [hunk 2]
```diff
@@ -23,8 +23,51 @@ SKIP_TESTS=false
...
+    changed="$(git diff --name-only "$TRAVIS_COMMIT" "$TRAVIS_BRANCH")"
+    if [ $? -ne 0 ]; then
+      # Fall back to git head
+      changed="$(git diff --name-only "$(git rev-parse HEAD)" "$TRAVIS_BRANCH")"
+      [ $? -ne 0 ] && return 0  # Give up. Just run everything.
+    fi
+
+    # Find the common prefix
+    prefix="$(echo "$changed" | \
+      # N: Do this for a pair of lines
```
Comment (tswast) [inline travis.sh]: 👍

**Pull request merged**